src/Microsoft.Build.Sql/VersionCheckTask.cs

```diff
@@ -0,0 +1,86 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Build.Framework;
+using NuGet.Common;
+using NuGet.Protocol;
+using NuGet.Protocol.Core.Types;
+using NuGet.Versioning;
+
+namespace Microsoft.Build.Sql;
+
+/// <summary>
+/// This task displays a message in the build log if there is a newer version of the package available on NuGet.org.
+/// </summary>
+public class VersionCheckTask : Microsoft.Build.Utilities.Task, ICancelableTask
+{
+    /// <summary>
+    /// Current version of the package.
+    /// </summary>
+    [Required]
+    public string Version { get; set; } = string.Empty;
+
+    private const string PackageName = "Microsoft.Build.Sql";
+    private const string NugetUrl = "https://api.nuget.org/v3/index.json";
+    private CancellationTokenSource _cancellationTokenSource = new();
+
+    public override bool Execute()
+    {
+        if (NuGetVersion.TryParse(Version, out NuGetVersion currentVersion) == false)
+        {
+            Log.LogWarning($"Invalid NuGetVersion format: {Version}");
+            return true;    // This task should not fail build even if the version check fails.
+        }
+
+        try
+        {
+            // Check for prerelease versions if current version is a prerelease version, otherwise check for stable versions only.
+            NuGetVersion latestVersion = GetLatestSdkVersionFromNuGet(currentVersion.IsPrerelease).Result;
```

Comment:
do we need to cache this? does it get called for each build?

Reply:
we have a primitive one day cache in our implementation

Reply:
We actually went for using the GitHub release API, much faster and more lightweight, and allowing more control of timeouts: https://github.com/rr-wfm/MSBuild.Sdk.SqlProj/blob/master/src/DacpacTool/VersionChecker.cs

Reply:
The Nuget API has a `SourceCacheContext` which I'll see if we can use. Yes, this is run for every build, the added overhead is one of the reasons .NET team gave for not implementing this.

@ErikEJ Thanks for the ideas. The Nuget library should just make an API call to this: https://learn.microsoft.com/en-us/nuget/api/package-base-address-resource but with Nuget typings. I'll do some benchmark to compare.

Reply:
In our experience the NuGet API frequently did not reply in the allowed time and we did not want to block the build process for many seconds.

Reply:
we should make sure it times out and doesn't add too much time to build. if it cannot get it ontime just cancel the request.

Reply:
Yes, as you can see in our implementation, we want max 2 seconds